docs: add initial plugin documentation

johnnygreco opened the pull request:
This PR adds some initial plugin documentation. Note that we expect this area to evolve a lot over the next couple months. 

I also added `analysis` to the code reference. 

Closes #107

nabinchha commented on src/data_designer/engine/dataset_builders/column_wise_builder.py:
```diff
@@ -171,6 +171,8 @@ def _run_cell_by_cell_generator(self, generator: ColumnGenerator) -> None:
             max_workers = generator.inference_parameters.max_parallel_requests
+        elif hasattr(generator.config, "max_parallel_requests"):
+            max_workers = generator.config.max_parallel_requests
         self._fan_out_with_threads(generator, max_workers=max_workers)
```
Nice! Yea without it the plugins were getting throttled to 4 from `MAX_CONCURRENCY_PER_NON_LLM_GENERATOR`. Just for the sake of consistency, I'd advocate pulling this out into a different PR since the PR title says `docs:`

nabinchha commented on docs/plugins/example.md:
```diff
@@ -0,0 +1,306 @@
+- Generic type `ColumnGenerator[IndexMultiplierColumnConfig]` connects the task to its config
+- `metadata()` describes your generator and its requirements
+- `generation_strategy` can be `FULL_COLUMN`, `ROW_WISE`, or `BATCH`
+- `required_resources` lists any required resources (models, artifact storage, etc.). This parameter will change in the future, so keeping it as `None` is safe for now.
```
One downside of keeping it as None is if your plugin depends on model registry. We rely on the dependency expressed here to know whether we need to run a health check for a model.

nabinchha commented on docs/plugins/example.md:
```diff
@@ -0,0 +1,306 @@
+!!! info "Understanding generation_strategy"
+    The `generation_strategy` specifies how the column generator will generate data.
+
+    - **`FULL_COLUMN`**: Generates the entire column at once
```
might be good to clarify, it's at the batch (buffer) level

nabinchha commented on docs/plugins/example.md:
```diff
@@ -0,0 +1,306 @@
+        - `generate` must take a `pd.DataFrame` as input and return a `pd.DataFrame`
+
+    - **`CELL_BY_CELL`**: Generates one cell at a time
+        - `generate` must take a `dict` as input and return a `dict`
```
might be good to mention dict provided keys with values from all previous columns

nabinchha commented on src/data_designer/config/analysis/column_statistics.py:
Just noticed the use of completion/prompt tokens here. In this PR[:](https://github.com/NVIDIA-NeMo/DataDesigner/pull/106) I'm updating this to input/output tokens. If this PR merges first, I'll update these.

nabinchha replied:
Or would that be a breaking change?

johnnygreco commented:
@nabinchha – addressed your comments in [378acb4](https://github.com/NVIDIA-NeMo/DataDesigner/pull/107/commits/378acb4ee54ff529e653ae3f9eff544b4eda1856)
